### PR TITLE
Rich redirects

### PR DESCRIPTION
**Pull request by gaaf**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [ ] Commits are split per component (core, individual modules, libs, utils, ...)
- [ ] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

If Kamailio sends a 301/302, the current dset is added as Contact header automatically. The feature in this PR will optionally extend this Contact header with more information. 

- The path vector for each contact can be added as a Contact-uri Route header (rfc 3327 conformant).
- The branch flags can be added as a Contact-header parameter (non-standard).

In addition, the get_redirects function is extended to optionally parse the new flags parameter. The Route header can be extracted from the r-uri in the script.

The new functionality must be enabled with module parameters, by default the behavior is unchanged.

This feature is based on a patchset by me as <alex@speakup.nl> published at https://lists.kamailio.org/pipermail/sr-dev/2011-August/012064.html

### CODE REVIEW

File: src/modules/uac_redirect/rd_funcs.c [hunk 1]
```diff
@@ -294,8 +295,16 @@ static int shmcontact2dset(struct sip_msg *req, struct sip_msg *sh_rpl,
...
+				if (flags_hdr_mode && scontacts[i]->flags && str2int(&(scontacts[i]->flags->body), &flags) == 0) {
+					if (flags_hdr_mode == 2)
+						flags |= bflags;
+				} else {
+					flags = bflags;
+				}
+			}
 			if(append_branch( 0, &scontacts[i]->uri, 0, 0, sqvalues[i],
-						bflags, 0, &_redirect_sruid.uid, 0,
+						flags, 0, &_redirect_sruid.uid, 0,
```
Comment (miconda) [inline src/modules/uac_redirect/rd_funcs.c]: As I got it, if flags parameter in contact is set and converted successfully to integer, even `flags_hdr_mode == 2`, the bflags is not taken in consideration.

Also I don't see how `flags_hdr_mode == 0` makes use of `bflags` like it was before for backward compatibility when `flags` param in contact is set.

File: src/core/dset.c [hunk 2]
```diff
@@ -595,6 +598,17 @@ static int print_contact_str(char **dest, str *uri, qvalue_t q, str *path, char
...
+
+	/* branch flags (not SIP standard conformant) */
+	buf.s = int2str(flags, &buf.len);
+	if (p + FLAGS_PARAM_LEN + buf.len > end) {
+		return -1;
+	}
+	memcpy(p, FLAGS_PARAM, FLAGS_PARAM_LEN);
+	p += FLAGS_PARAM_LEN;
+	memcpy(p, buf.s, buf.len);
+	p += buf.len;
```
Comment (miconda) [inline src/core/dset.c]: This should be made somehow optional, not being standard, as well as eventually do not add the flags parameter if the value is 0 (not set).

Comment (gaaf) [reply src/core/dset.c]: Patch 4 makes this optional.

Comment (miconda) [reply src/core/dset.c]: Seems that follow up patches add the option. For the future, when doing the PR merge the patches done to the same component in one, makes the review simpler as one does it in the order they are listed and later discovers is no longer same code, being again changed.

File: src/modules/uac_redirect/rd_funcs.c (hunk 1 above)
Comment (gaaf) [reply src/modules/uac_redirect/rd_funcs.c]: 1. If conversion is successful, flags are OR'ed with bflags when mode == 2.

There is something wonky here, looks like an incorrectly resolved merge/rebase. I'll fix this up.

2. Good catch, i'll fix this.

**Comment by gaaf** [comment]
> @gaaf - any updates on this?

I still have this on my todo list, just haven't found the time to fix it yet. Sorry for the delay.

**Comment by miconda** [comment]
@gaaf - soft reminder just in case you can address this one and eventually have it merged.

File: src/modules/uac_redirect/rd_funcs.c (hunk 1 above)
Comment (gaaf) [reply src/modules/uac_redirect/rd_funcs.c]: Both issues were caused by the same broken merge/rebase. Solved in my new patchset.

**Comment by gaaf** [comment]
github's broken commit sorting messes up the order of the patches, better review it as a whole.

**Comment by gaaf** [comment]
Worked around github's brokenness, commits are now in the correct order.

**Comment by miconda** [comment]
Thanks! Looks ok after a quick review, I will go over one more try and then merge if nothing discovered.

**Comment by henningw** [comment]
@miconda - i merged it into master, if there is something else to be found we can fix it there.

**Pull request merged**